Max number of columns in an output file

guilpier-code opened the pull request:
The current PR was first an attempt the issue #927 

It now contains a preparatory refactoring work before addressing the real problem raised by #927.
Issue #927 is addressed in #1172. 

**The original purpose was** : 
The purpose is to compute more precisely the max number of columns that a output report can have than it is today.
There is only one container that holds the results (see `SurveyResults::maxVariables`), and it has to be sized at the max number of columns a report can have.
As the synthesis output reports have more columns than the year-by-year reports, we build this max number from their number of columns.
The goal here can be summarized as follows : 
```C++
void SurveyResults::initializeMaxVariables(const Data::Study& s)
{
     // Getting the any report's max number of columns     
     maxVariables = s.parameters.variablesPrintInfo.getMaxColumnsCount();
     logs.debug() << "  (for " << maxVariables << " columns)";
}
```

**Tests** : 
To validate this work, tests are passed : on [this repo](https://github.com/AntaresSimulatorTeam/Antares_Simulator_Tests_NR), in directory **valid-filter-outputs**.

**Eventually** : 
This first attempt is a half-failure.
It gave the opportunity to make some refactoring and cleaning but the purpose of this issue could not be reached.
For further developments allowing to thoroughly reach the target, see [this comment](https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/1159#issuecomment-1429581033).

guilpier-code committed:
max nb columns for print : simplify class AllVariablesPrintInfo

guilpier-code committed:
max nb columns for print : simplify class AllVariablesPrintInfo (part 2)

guilpier-code commented on src/libs/antares/study/variable-print-info.cpp:
```diff
@@ -230,8 +185,8 @@ bool AllVariablesPrintInfo::isPrinted(std::string var_name) const
+        v->enablePrint(b);
 }
 
 void AllVariablesPrintInfo::computeMaxColumnsCountInReports()
```
About **computeMaxColumnsCountInReports()** : 
This computation is strange.
There are 2 degrees of freedom : **FileLevel** (value, id, details, ...) and **DataLevel** (area, links, ...).
But here we make just one loop. Should not we make 2 nested loops ?

guilpier-code replied:
Done in #1172

guilpier-code commented on src/libs/antares/study/variable-print-info.cpp:
```diff
@@ -186,42 +171,12 @@ void AllVariablesPrintInfo::prepareForSimulation(bool userSelection,
-            return (*it_info)->isPrinted();
-        }
-    }
+    auto has_name = [&](VariablePrintInfo* v) { return v->name() == var_name; };
```
About **isPrinted** and **setPrintStatus** : 
These functions let think that **allVarsPrintInfo** should be a **std::map** instead of a **std::vector**.
If so, we could have : 
```C++
bool AllVariablesPrintInfo::isPrinted(std::string var_name) const
{
    return VariablePrintInfo[var_name]->isPrinted();
}
```
To be thought of (it may not be that easy)

guilpier-code commented on src/libs/antares/study/variable-print-info.cpp:
```diff
@@ -86,12 +86,6 @@ void variablePrintInfoCollector::add(const AnyString& name,
 // ============================================================
```
About **clear()** function : 
Currently, we need a **clear()** in class **AllVariablesPrintInfo** because the vector of **VariablePrintInfo** expects pointers to **VariablePrintInfo**.
Instead of : 
```C++
  std::vector<VariablePrintInfo*> allVarsPrintInfo;
```
we should have : 
```C++
  std::vector<VariablePrintInfo> allVarsPrintInfo;
```
This would allow to remove the **clear()** function and the **destructor** to be a **default** one.

guilpier-code committed:
max nb columns for print : move all computations in one place (SurveyResults's constructor)

guilpier-code commented:
As said the current PR's description, we could not reach the purpose of this PR.
The reason is that the operation of visiting all output variables (`RetrieveVariableList(collector)`) is static : it is made at compilation time.
When compiling, there is no way to know about the size required by the **dynamic** output variables 
(`columnCount = Category::dynamicColumns`).

A possible solution : 
- Today, **variablePrintInfoCollector** is built at compilation time and configured at runtime. But more importantly, besides being used to handle the print status of each variables, it is also required to collect the max size of each output variable, which it cannot do accurately (because of dynamic columns variables). So it should not be used to collect max size of variables.
- As we have to collect max size of each output variables at runtime, we could do what is done for other reasons elsewhere : defining and calling another function to traverse the static list of output variables at runtime. The function's prototype could be `void getMaxColumnsNumbe(SizeCollector &)`.
- A said in https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/1159#discussion_r1105707798, the today's place to collect the max number of columns is not suitable. A possible suitable place where to instantiate the **SizeCollector** and run the **getMaxColumnsNumbe(...)** could be the `void ISimulation<Impl>::run()` method, right after the call of **variables.initializeFromStudy(...)**.

guilpier-code commented on src/solver/variable/surveyresults/surveyresults.cpp:
```diff
@@ -504,10 +504,16 @@ static inline void WriteIndexHeaderToFileDescriptor(int precisionLevel,
-uint initializeMaxVariables(uint maxVars, const Data::StudyRuntimeInfos* runtime)
+void SurveyResults::initializeMaxVariables(const Data::Study& s)
 {
+    const auto* runtime = s.runtime;
```
Note that getting the max column count here is inefficient : located here, this operation is done for each yearly results and for synthesis results. But it could be done only once.
We should do it elsewhere : see https://github.com/AntaresSimulatorTeam/Antares_Simulator/pull/1159#issuecomment-1429581033

flomnes replied:
OK I identified a few steps

1. Compute the matrix size required to store all results
2. Allocate memory
3. Write to that memory
4. Dump that memory into files
5. Deallocate memory

What interests us here is step 1. Currently, it is done both statically and dynamically.

Static: Number of variables with fixed number of columns (mostly 1, sometimes more)
Dynamic: Number of clusters * number of variables by cluster (hard-coded)

Then we take the max of all this to make sure that every result fits (details-xxx.txt, values-xxx.txt, etc.).

For every file, there is a category id (de, de_res, etc.) that's stored in attribute `categoryFileLevel`. Example
```cpp
// src/solver/variable/economy/profitByPlant.h
categoryFileLevel = ResultsType::categoryFile & (Category::de),
// src/solver/variable/economy/dispatchable-generation-margin.h
categoryFileLevel = ResultsType::categoryFile & (Category::id | Category::va),
```

For every category, we need to compute the width & height of the associated matrix for yby & synthesis files.

Why not attach a function to every variable ? It could return 1 (most cases), another constant (thermal/renewable clusters), or a number depending on the study ("this variable produces one variable per thermal cluster")
```cpp
categoryFileLevel = ResultsType::categoryFile & (Category::de),
columnCalculator = [](const Study& study) { // return the number of thermal clusters };
```

Or, if you don't like lambda functions, we could have a small library of functions (return a constant, return the number of thermal clusters, return the number of inequality binding constraints).

guilpier-code replied:
- Currently, the computation of max number of columns is done before the clusters or binding constraints are loaded.
So in the state of the current code, it would not work. 
That said, we could change the order in the data load.
- **Question** : we will have to traverse all output variables anyway whether we call a classic method or a lambda function, right ? How do you think we could do ?

flomnes replied:
> Currently, the computation of max number of columns is done before the clusters or binding constraints are loaded.

No, see function `initializeMaxVariables` & constructor of class `SurveyResults`

guilpier-code replied:
> No, see function `initializeMaxVariables` & constructor of class `SurveyResults`

`initializeMaxVariables` is based on information collected before the clusters and BCs are loaded.
`initializeMaxVariables` only gets their value.

guilpier-code committed:
max nb columns for print : computing an accurate max number of columns (based on synthesis reports)

Note that synthesis reports contain more columns than year by year reports.
So the computed max number of columns is actually the max number of columns in a synthesis report

guilpier-code committed:
max nb columns for print : more cleaning (#1180)

* max nb columns for print : code improvement

Improvements :
- At reading the output variable print status (parameters.cpp)
- Small changes inside classes of variable-print-info.h/.cpp
- Removing function SurveyResults::initializeMaxVariables(...), which has shrunk until it does one function call.

* max nb columns for print : in class AllVariablesPrintInfo, changing the std::vector into a std::map

* max nb columns for print : renaming in output variable BC MARG COST

* max nb columns for print : fix a bug

* max nb columns for print : fix other bugs

* max nb columns for print : refresh names lists each time a print status is changed

* max nb columns for print : removing useless isEmpty()

Indeed, allVarsPrintInfo can never be empty, has it is build from code built at compile time

* max nb columns for print : try fix a compilation failure on Linux

* max nb columns for print : change for more readable manner to traverse std::map

* max nb columns for print : corrections after review

* max nb columns for print : more corrections after review

* max nb columns for print : amend previous commit (causes compilation error on Linux)

* max nb columns for print : remove lists that have to be updated in real-time

* Use enum value in template param (#1193)

* max nb columns for print : adding member function numberOfEnabledVariables() to class AllVariablesPrintInfo

---------

Co-authored-by: Florian Omnès <florian.omnes@rte-france.com>

flomnes commented on src/libs/antares/study/variable-print-info.cpp:
```diff
@@ -186,42 +171,12 @@ void AllVariablesPrintInfo::prepareForSimulation(bool userSelection,
-        }
-    }
+    auto has_name = [&](VariablePrintInfo* v) { return v->name() == var_name; };
+    auto it = std::find_if(allVarsPrintInfo.begin(), allVarsPrintInfo.end(), has_name);
+    if (it != allVarsPrintInfo.end())
+        return (*it)->isPrinted();
```
Using [std::any_of](https://en.cppreference.com/w/cpp/algorithm/all_any_none_of) will save you a few lines of code

guilpier-code replied:
Due to changes, remark not relevant anymore

flomnes commented:
Please resolve conflicts

guilpier-code committed:
Merge remote-tracking branch 'remotes/origin/develop' into fix/max-nb-colunms-for-print

flomnes commented on src/libs/antares/study/variable-print-info.cpp:
```diff
@@ -76,161 +79,141 @@ variablePrintInfoCollector::variablePrintInfoCollector(AllVariablesPrintInfo* al
+    std::string upperCaseName = to_uppercase(name);
+    if (not exists(upperCaseName))
+    {
+        index_to_name[(unsigned int)allVarsPrintInfo.size()] = upperCaseName;
+        allVarsPrintInfo.insert(std::pair<std::string, VariablePrintInfo>(upperCaseName, v));
```
Use `try_emplace`, see https://www.fluentcpp.com/2018/12/11/overview-of-std-map-insertion-emplacement-methods-in-cpp17/

guilpier-code replied:
Done

flomnes commented on src/libs/antares/study/variable-print-info.cpp:
```diff
@@ -239,56 +222,58 @@ void AllVariablesPrintInfo::computeMaxColumnsCountInReports()
+            for (auto& [name, variable] : allVarsPrintInfo)
+            {
+                if (variable.isPrinted() &&
+                    variable.getFileLevel() & CFileLevel &&
+                    variable.getDataLevel() & CDataLevel)
```
```suggestion
                    variable.isFileLevel(CFileLevel) &&
                    variable.isDataLevel(CDataLevel))
```
Apply the `&` operator inside the method instead. Looks a bit more clear

guilpier-code replied:
Done

flomnes commented on src/libs/antares/study/variable-print-info.cpp:
```diff
@@ -239,56 +222,58 @@ void AllVariablesPrintInfo::computeMaxColumnsCountInReports()
+    for (auto& [name, variable] : allVarsPrintInfo)
     {
-        if ((*it_info)->isPrinted() && (*it_info)->getDataLevel() == Category::area)
+        if (variable.isPrinted() && variable.getDataLevel() == Category::area)
```
Use `std::count_if`

guilpier-code replied:
Done

flomnes commented on src/libs/antares/study/variable-print-info.cpp:
```diff
@@ -239,56 +222,58 @@ void AllVariablesPrintInfo::computeMaxColumnsCountInReports()
+    for (auto& [name, variable] : allVarsPrintInfo)
     {
-        if ((*it_info)->isPrinted() && (*it_info)->getDataLevel() == Category::link)
+        if (variable.isPrinted() && variable.getDataLevel() == Category::link)
```
Use `std::count_if`

guilpier-code replied:
Done

flomnes commented on src/libs/antares/study/variable-print-info.h:
```diff
@@ -44,38 +44,35 @@ namespace Data
-    }
+    void setMaxColumns(uint maxColumnsNumber);
+
+    uint getDataLevel() { return dataLevel_; }
+    uint getFileLevel() { return fileLevel_; }
```
```suggestion
    bool isDataLevel(uint dataLevel) const { return dataLevel_ & dataLevel; }
    bool isFileLevel(uint fileLevel) const { return fileLevel_ & fileLevel; }
```

guilpier-code replied:
Done

flomnes commented on src/solver/variable/surveyresults/surveyresults.cpp:
```diff
@@ -489,54 +489,21 @@ static inline void WriteIndexHeaderToFileDescriptor(int precisionLevel,
 // TOFIX - MBO 02/06/2014 nombre de colonnes fonction du nombre de variables
-SurveyResults::SurveyResults(uint maxVars,
-                             const Data::Study& s,
+SurveyResults::SurveyResults(const Data::Study& s,
```
You may now remove this comment

guilpier-code replied:
Done

flomnes commented on src/libs/antares/study/variable-print-info.cpp:
```diff
@@ -76,161 +79,141 @@ variablePrintInfoCollector::variablePrintInfoCollector(AllVariablesPrintInfo* al
 
-void AllVariablesPrintInfo::prepareForSimulation(bool userSelection,
-                                                 const std::vector<std::string>& excluded_vars)
+std::string AllVariablesPrintInfo::name_of(unsigned int index) const
```
```suggestion
[[deprecated("Only needed by the GUI, to be removed")]]
std::string AllVariablesPrintInfo::name_of(unsigned int index) const
```

guilpier-code replied:
Done

flomnes commented on src/libs/antares/study/variable-print-info.h:
```diff
@@ -97,30 +94,30 @@ class AllVariablesPrintInfo
-    void add(VariablePrintInfo* v);
+    void add(std::string name, VariablePrintInfo v);
     void clear();
-    VariablePrintInfo* operator[](uint i) const;
+    VariablePrintInfo& operator[](uint i);
     size_t size() const;
-    bool isEmpty() const;
+    bool exists(std::string name);
```
Use `const std::string&` or `std::string_view`

guilpier-code replied:
I tried to replace **std::string** by **std::string_view** or  **const std::string&** in arguments of member functions : 
- **add(...)**
- **exists(...)**
- **setPrintStatus(...)**
- **isPrinted(...)**

but they all call **to_uppercase(std::string& str)** internally, that makes changes on the **std::string**.
So I leave the code as it is regarding this.

flomnes commented on src/libs/antares/study/variable-print-info.h:
```diff
@@ -132,27 +129,32 @@ class AllVariablesPrintInfo
+    void reverseAll();
+
+    unsigned int numberOfEnabledVariables();
+    std::vector<std::string> namesOfEnabledVariables();
+    std::vector<std::string> namesOfDisabledVariables();
```
Avoid copies (`std::string_view` are non-owning references to char ranges)
```suggestion
    std::vector<std::string_view> namesOfEnabledVariables();
    std::vector<std::string_view> namesOfDisabledVariables();
```

guilpier-code replied:
The returned **std::vector\<std::string\>** is an **rvalue**, meaning a temporary value.
Do we really want to make it a **std::vector\<std::string_view\>** ?

flomnes reviewed (changes requested):
Minor changes required. LGTM

guilpier-code committed:
max nb columns for print : corrections after review

guilpier-code committed:
max nb columns for print : removing unnecessary member functions

guilpier-code commented:
As you suggested, I removed : 
- **VariablePrintInfo::hasDataLevel(...)**
- **VariablePrintInfo::hasFileLevel(...)**
